File: backend/app/web/shared/speech.py
"""Text destined to be spoken aloud — a port of `web_version/talbotiq-platform/shared/speech.ts`.

Everything here exists because a written question and a spoken one are not the same
artefact. Markdown becomes noise ("asterisk asterisk bold"), a numbered list gets its
numbers read out, and an em dash reads as a machine wrote it.

Like `invite_email.py`, this module is imported by the React client as well as the
Express server, so the port introduces a second implementation of shared logic. Unlike
the email renderer, nothing here is byte-compared between the two: the output is a
prompt, not a document, and a small divergence changes phrasing rather than correctness.
Worth knowing rather than guarding.
"""

from __future__ import annotations

import re

TIME_GREETINGS = {
    "morning": "Good morning",
    "afternoon": "Good afternoon",
    "evening": "Good evening",
}

# How much résumé the avatar is given as background. Enough to sound informed, bounded
# because it rides in every conversation's context.
MAX_BACKGROUND_CHARS = 1500

_MARKDOWN = re.compile(r"[*_`#>]+")
_LIST_PREFIX = re.compile(r"^\s*(?:\d+[.)]|[-–—•])\s+", re.MULTILINE)
_DASHES = re.compile(r"\s*[—–]\s*|\s+-\s+")
_WHITESPACE = re.compile(r"\s+")

# Every English variant the recogniser may settle on. Naming them all — rather than a
# single "en-US" — is what stops an Indian-accented answer being scored as a different
# language entirely: with no hint at all the recogniser is free to decide the audio is
# Hindi and transcribe Devanagari into an English-only interview.
ENGLISH_VARIANTS = ("en-IN", "en-US", "en-GB", "en-AU")

# Ported from the Express implementation (server/services/voice.ts:108-121), which these
# were dropped from in the FastAPI port.
_ADAPTATION_TOKEN = re.compile(r"[A-Za-z][A-Za-z0-9+#.]*[A-Za-z0-9+#]")
_ADAPTATION_ACRONYM = re.compile(r"^[A-Z]{2,}[0-9+#]*$")
_ADAPTATION_CAMEL = re.compile(r"[a-z][A-Z]")
_ADAPTATION_DIGIT = re.compile(r"[0-9+#]")
_ADAPTATION_DOTTED = re.compile(r"\w\.\w")
_ADAPTATION_PROPER = re.compile(r"[A-Z][a-z0-9]+(?:[ -][A-Z][a-z0-9]+)+")
# A single capitalised word MID-SENTENCE: "Redis", "Grafana", "Kafka". The Express
# heuristic missed exactly these — they are neither acronyms nor multi-word — which is
# why candidates heard "Redis" come back as "reduce". Sentence-initial words are skipped
# so ordinary openers ("How", "Describe") never become hints.
_ADAPTATION_SENTENCE = re.compile(r"(?<=[.!?\n])\s*|^\s*")
_ADAPTATION_PROPER_WORD = re.compile(r"^[A-Z][a-z0-9]{2,}$")

# Google caps the hint list; 32 is what the Express relay sent.
MAX_ADAPTATION_PHRASES = 32


def greeting_word(time_of_day: str | None) -> str:
    return TIME_GREETINGS.get(time_of_day or "", "Hello")


def transcription_languages(language: str | None) -> list[str]:
    """The language hints for the recogniser.

    An English interview is hinted with every English variant, because a candidate's
    accent is not a different language. Anything else is passed through untouched.
    """
    value = (language or "").strip().lower()
    if not value or value.startswith("en") or value.startswith("english"):
        return list(ENGLISH_VARIANTS)
    return [language.strip()]


def adaptation_phrases(role: str | None, questions: list[str]) -> list[str]:
    """Domain terms to bias the recogniser towards.

    Without these, "Redis" comes back as "reduce" and "Grafana" as "nature" — the
    recogniser has no reason to prefer a technical term over a common English word that
    sounds like it. The question script is the best available source of the vocabulary a
    given interview will actually contain.
    """
    out: list[str] = []
    seen: set[str] = set()

    def add(phrase: str) -> None:
        if phrase and phrase not in seen:
            seen.add(phrase)
            out.append(phrase)

    if (role or "").strip():
        add(role.strip()[:60])

    text = "\n".join(questions)
    for match in _ADAPTATION_TOKEN.finditer(text):
        word = match.group(0)
        if not 2 <= len(word) <= 40:
            continue
        if (
            _ADAPTATION_ACRONYM.match(word)
            or _ADAPTATION_CAMEL.search(word)
            or _ADAPTATION_DIGIT.search(word)
            or _ADAPTATION_DOTTED.search(word)
        ):
            add(word)

    for match in _ADAPTATION_PROPER.finditer(text):
        add(match.group(0)[:60])

    for sentence in _ADAPTATION_SENTENCE.split(text):
        if not sentence:
            continue
        # Skip the first word: it is capitalised by grammar, not because it is a name.
        for word in sentence.split()[1:]:
            stripped = word.strip(".,;:!?()[]\"'")
            if _ADAPTATION_PROPER_WORD.match(stripped):
                add(stripped)

    return out[:MAX_ADAPTATION_PHRASES]


def strip_for_speech(text: str) -> str:
    """Reduce written text to something that reads naturally aloud.

    Markdown, list markers and dashes all have visual meaning and no spoken one — a
    text-to-speech engine either voices them literally or pauses oddly around them.
    """
    value = _LIST_PREFIX.sub("", text or "")
    value = _MARKDOWN.sub("", value)
    value = _DASHES.sub(", ", value)
    value = re.sub(r",\s*,", ",", value)
    return _WHITESPACE.sub(" ", value).strip()


SPOKEN_STYLE_RULES = (
    "SPEAKING STYLE: You are speaking out loud, not writing. Use contractions and short, "
    "natural sentences. Never read out formatting, numbers of questions, or lists. Never "
    "use em dashes. Sound warm and genuinely interested, never templated or corporate."
)

VARIED_THANKS_RULE = (
    "After each answer, give ONE short, varied, genuine acknowledgment before the next "
    "question — never the same phrase twice, and never a critical one."
)

# Turn-taking is guessed from silence, and silence is ambiguous: a candidate gathering
# their next point looks exactly like a candidate who has finished. Guessing wrong in the
# "finished" direction silently costs them the rest of their answer, and they cannot get
# it back — the interview has already moved on. Asking costs one short line.
#
# The check is phrased as a question so it is never mistaken for a wrap-up, and the added
# detail lands under the same question for scoring: a confirmation does not match any
# planned question, so the server records it as an acknowledgment and leaves the question
# cursor where it is (app/web/services/avatar_transcript.py).
CONFIRM_BEFORE_ADVANCING_RULE = (
    "BEFORE MOVING ON — never assume an answer is finished just because they paused, but "
    "use this check SPARINGLY: when they give a complete, substantial answer, do NOT ask "
    "anything — acknowledge briefly and go straight to the next question. ONLY when their "
    "answer is very short (a sentence or two), cut off mid-thought, or clearly unfinished, "
    "ask ONE short check such as \"Is that everything, or is there anything you'd like to "
    "add?\" and WAIT. Vary the wording, and never use the check twice on the same "
    "question. Anything they add is part of that same answer, so let them finish it. This "
    "applies to EVERY question, including the last one."
)


def resume_addendum(asked: list[str]) -> str:
    """Instructions for a session that is picking up after a dropped connection.

    A fresh Live session has no memory of the call it is replacing, so without this the
    interviewer greets the candidate a second time and works through the script from the
    top — asking questions they have already answered. Naming what was covered is what
    makes the continuation seamless; the Express relay carried the same clause
    (server/services/voice.ts:424-426) and the port dropped it.
    """
    covered = "\n".join(f"- {strip_for_speech(question)}" for question in asked)
    already = (
        f"\n\nYou have ALREADY asked and received answers to these, so do NOT ask them "
        f"again:\n{covered}"
        if asked
        else ""
    )
    return (
        "RESUMING: the connection dropped briefly and has just been restored. This is the "
        "SAME interview continuing, not a new one. Do NOT greet the candidate again, do "
        "NOT introduce yourself again, and do NOT start over. Say one short line to the "
        "effect that you are back and briefly apologise for the interruption, then "
        "continue with the next planned question you had not yet covered."
        f"{already}"
    )


def default_interviewer_persona(
    candidate_name: str | None = None, ai_name: str | None = None
) -> str:
    who = (candidate_name or "").strip() or "the candidate"
    me = (ai_name or "").strip() or "Alex"
    return (
        f"You are {me}, a Senior Talent Specialist at TalbotIQ conducting a screening "
        f"interview with {who}. You are warm, personable and encouraging. You put people "
        "at ease and sound genuinely interested in their answers."
    )


def avatar_interview_context(
    *,
    persona_text: str | None = None,
    candidate_name: str | None = None,
    ai_name: str | None = None,
    questions: list[str],
    time_of_day: str | None = None,
    resume_text: str | None = None,
    confirm_before_advancing: bool = False,
    closing_can_leave: bool = False,
) -> str:
    """The avatar's full instructions for one interview.

    The strict-script rules are the load-bearing part. An avatar left to improvise asks
    different questions of different candidates, which makes the scores incomparable and
    the screen indefensible — so it is told, explicitly and more than once, to ask exactly
    these questions in exactly this order and invent nothing.

    The résumé is included as BACKGROUND only, with the same prohibition attached: it is
    there so the avatar sounds informed when it acknowledges an answer, not so it can
    think of new questions.
    """
    who = (candidate_name or "").strip() or "the candidate"
    named = bool((candidate_name or "").strip())
    persona = (persona_text or "").strip() or default_interviewer_persona(
        candidate_name, ai_name
    )
    numbered = "\n".join(
        f"{index + 1}. {strip_for_speech(question)}"
        for index, question in enumerate(questions)
    )

    sections = [persona]

    if (resume_text or "").strip():
        sections.append(
            f"CANDIDATE BACKGROUND, from {who}'s résumé. Use it to sound informed and to "
            "personalise your brief acknowledgments naturally (e.g. referencing their "
            "experience), but NEVER to add, change, or skip scripted questions:\n"
            f"{resume_text.strip()[:MAX_BACKGROUND_CHARS]}"
        )

    sections += [
        SPOKEN_STYLE_RULES,
        f"""FLOW:
1. Open with a brief "{greeting_word(time_of_day)}" greeting and warmly welcome {who}{' by name' if named else ''}. Add one short reassuring line about how this will go, then ask if they're ready to begin, and wait.
2. If they clearly say yes, begin. If they're unsure or nervous, reassure them in one short line and ask again; only start on a clear yes.
3. Ask the questions below IN ORDER, one at a time, phrased exactly as written. Wait for {who} to completely finish each answer — never interrupt. {VARIED_THANKS_RULE}
4. Only AFTER the final question is answered, close warmly: thank them sincerely, tell them that's everything and they're all done, that the team will be in touch about next steps, and wish them a great rest of their day.{" Finally, tell them they're free to leave now and that the session will close by itself shortly." if closing_can_leave else ""}""",
        *([CONFIRM_BEFORE_ADVANCING_RULE] if confirm_before_advancing else []),
        "THE QUESTIONS, IN ORDER. Ask every one, exactly as written; never say their "
        f"numbers aloud:\n{numbered}",
        f"STRICT RULES: Ask ONLY these questions. Do NOT invent, add, skip, reorder, or "
        f"rephrase any question, and never ask follow-ups that are not in the list"
        + (
            " — the one exception is the short \"anything to add?\" check above, which is "
            "not a new question and does not replace or count as one of the planned "
            "questions"
            if confirm_before_advancing
            else ""
        )
        + f". No small talk beyond the opening. If {who} goes off-topic or asks you "
        "questions, politely acknowledge in one short line and steer straight back to the "
        "next planned question. Cover ALL the questions, then close — never finish early "
        "and never add questions of your own.",
    ]

    return "\n\n".join(sections)


def avatar_greeting_text(
    *,
    custom: str | None = None,
    candidate_name: str | None = None,
    ai_name: str | None = None,
    time_of_day: str | None = None,
) -> str:
    """The avatar's first words. A recruiter's own greeting wins."""
    if (custom or "").strip():
        return strip_for_speech(custom)

    hello = greeting_word(time_of_day)
    name = (candidate_name or "").strip()
    me = (ai_name or "").strip()
    intro = (
        f"I'm {me}, and I'm really looking forward to our chat"
        if me
        else "I'm really looking forward to our chat"
    )
    who = f" {name}," if name else ","

    return (
        f"{hello}{who} welcome, and thanks so much for making the time today. {intro}, so "
        "just relax and answer naturally. Are you ready to begin?"
    )

File: web_version/talbotiq-platform/shared/speech.ts
import type { TimeOfDay } from './types'

/**
 * The ONE voice & persona standard — shared by every surface that speaks:
 *
 *   • Voice Interview  (server/services/voice.ts — Gemini Live system instruction)
 *   • Tavus avatar     (server/services/tavusServer.ts — candidate conversations,
 *                       and src/pages/SetupPage.tsx — recruiter-run sessions)
 *   • Mimic Guide      (src/lib/guideSpeech.ts — text prepared for TTS)
 *
 * Two levers live here:
 *   1. stripForSpeech() — nothing markdown/formatted is ever read aloud.
 *   2. The persona standard — warm, polite, contractions, varied phrasing,
 *      greets, thanks after every answer — expressed once and reused, so all
 *      three surfaces sound like the same human product and are tuned in
 *      ONE place.
 *
 * Plain runtime TypeScript (no imports beyond shared types) so both the Vite
 * client and the Express server can import it.
 */

/* ─── Lever 1: never read formatting aloud ──────────────────────────────── */

/**
 * Convert model/authored text into clean, natural SPOKEN sentences:
 * markdown, bullets, headings, code, links, tables, emoji and stray symbols
 * are removed or unwrapped — "asterisk asterisk" is never spoken.
 */
export function stripForSpeech(text: string): string {
  return (
    text
      // fenced code blocks first (keep a spoken hint rather than reading code)
      .replace(/```[\s\S]*?```/g, ' ')
      // HTML tags (incl. <details> wrappers already removed by callers)
      .replace(/<\/?[a-z][^>]*>/gi, ' ')
      // links/images: keep the label, drop the URL
      .replace(/!?\[([^\]]*)\]\([^)]*\)/g, '$1')
      // bare URLs
      .replace(/https?:\/\/\S+/g, ' ')
      // bold / italic / strikethrough / inline code — unwrap, keep the words
      .replace(/(\*\*\*|\*\*|\*|___|__|_|~~|`)(.*?)\1/g, '$2')
      // headings, blockquotes, list bullets at line starts
      .replace(/^[ \t]*(?:#{1,6}[ \t]*|>+[ \t]*|[-*+][ \t]+|\d+[.)][ \t]+)/gm, '')
      // table pipes / separators
      .replace(/^\|.*\|$/gm, (row) => row.replace(/\|/g, ' ').replace(/-{3,}/g, ' '))
      // emoji & pictographs (TTS engines skip or misread them)
      .replace(/[\u{1F300}-\u{1FAFF}\u{2600}-\u{27BF}\u{FE0F}\u{200D}]/gu, ' ')
      // leftover markdown noise
      .replace(/[*_#`~|]/g, ' ')
      // em/en dashes read poorly on some engines — a comma phrases naturally
      .replace(/\s*[—–]\s*|\s+-\s+/g, ', ')
      .replace(/\s+/g, ' ')
      .trim()
  )
}

/* ─── Lever 1b: the spoken voice must follow the CONTENT's language ──────── */

/** BCP-47 locale for every guide language (STT `recognition.lang` + TTS). */
export const SPEECH_LOCALES: Record<string, string> = {
  en: 'en-US', hi: 'hi-IN', mr: 'mr-IN', ta: 'ta-IN', te: 'te-IN', kn: 'kn-IN',
  ml: 'ml-IN', gu: 'gu-IN', pa: 'pa-IN', bn: 'bn-IN', ur: 'ur-PK', zh: 'zh-CN',
  'zh-TW': 'zh-TW', ja: 'ja-JP', ko: 'ko-KR', ar: 'ar-SA', fa: 'fa-IR',
  tr: 'tr-TR', ru: 'ru-RU', uk: 'uk-UA', pl: 'pl-PL', cs: 'cs-CZ', sk: 'sk-SK',
  ro: 'ro-RO', hu: 'hu-HU', de: 'de-DE', fr: 'fr-FR', es: 'es-ES', pt: 'pt-BR',
  it: 'it-IT', nl: 'nl-NL', sv: 'sv-SE', no: 'nb-NO', da: 'da-DK', fi: 'fi-FI',
  el: 'el-GR', he: 'he-IL', id: 'id-ID', ms: 'ms-MY', th: 'th-TH', vi: 'vi-VN',
  fil: 'fil-PH', sw: 'sw-KE', af: 'af-ZA', am: 'am-ET', az: 'az-AZ', be: 'be-BY',
  bg: 'bg-BG', bs: 'bs-BA', ca: 'ca-ES', hr: 'hr-HR', lt: 'lt-LT', lv: 'lv-LV',
  sr: 'sr-RS', sl: 'sl-SI',
}

/** Languages whose script is non-Latin (a wrong-language voice reading these
 *  skips the glyphs entirely, so they must never fall back to an English voice). */
export const NON_LATIN_LANGS = new Set<string>([
  'hi', 'mr', 'ta', 'te', 'kn', 'ml', 'bn', 'gu', 'pa', 'ur', 'zh', 'zh-TW',
  'ja', 'ko', 'ar', 'fa', 'ru', 'uk', 'be', 'bg', 'sr', 'el', 'he', 'th', 'am',
])

/** Unicode-script → locale rules, checked against the TEXT being spoken. A
 *  shared alphabet resolves via the user's selected language (e.g. Devanagari
 *  is Hindi unless Marathi is selected; Han is Chinese unless kana is present). */
const SCRIPT_RULES: Array<{ re: RegExp; locale: (prefer: string, text: string) => string }> = [
  { re: /[\u0C00-\u0C7F]/g, locale: () => 'te-IN' }, // Telugu
  { re: /[\u0C80-\u0CFF]/g, locale: () => 'kn-IN' }, // Kannada
  { re: /[\u0B80-\u0BFF]/g, locale: () => 'ta-IN' }, // Tamil
  { re: /[\u0D00-\u0D7F]/g, locale: () => 'ml-IN' }, // Malayalam
  { re: /[\u0900-\u097F]/g, locale: (p) => (p === 'mr' ? 'mr-IN' : 'hi-IN') }, // Devanagari
  { re: /[\u0980-\u09FF]/g, locale: () => 'bn-IN' }, // Bengali
  { re: /[\u0A00-\u0A7F]/g, locale: () => 'pa-IN' }, // Gurmukhi
  { re: /[\u0A80-\u0AFF]/g, locale: () => 'gu-IN' }, // Gujarati
  { re: /[\u0600-\u06FF\u0750-\u077F]/g, locale: (p) => (p === 'ur' ? 'ur-PK' : p === 'fa' ? 'fa-IR' : 'ar-SA') }, // Arabic script
  { re: /[\u0590-\u05FF]/g, locale: () => 'he-IL' }, // Hebrew
  { re: /[\u0400-\u04FF]/g, locale: (p) => (['ru', 'uk', 'be', 'bg', 'sr'].includes(p) ? SPEECH_LOCALES[p] : 'ru-RU') }, // Cyrillic
  { re: /[\u0370-\u03FF]/g, locale: () => 'el-GR' }, // Greek
  { re: /[\u0E00-\u0E7F]/g, locale: () => 'th-TH' }, // Thai
  { re: /[\u1200-\u137F]/g, locale: () => 'am-ET' }, // Ethiopic
  { re: /[\uAC00-\uD7AF\u1100-\u11FF\u3130-\u318F]/g, locale: () => 'ko-KR' }, // Hangul
  { re: /[\u3040-\u30FF]/g, locale: () => 'ja-JP' }, // Kana
  { re: /[\u4E00-\u9FFF]/g, locale: (p, text) => (/[\u3040-\u30FF]/.test(text) ? 'ja-JP' : p === 'zh-TW' ? 'zh-TW' : 'zh-CN') }, // Han
]

/**
 * Pick the TTS locale from the CONTENT being spoken — never from a hardcoded
 * default. The dominant non-Latin script in the text decides the language
 * (so a Telugu answer speaks Telugu even while the selector shows English);
 * `preferredLang` only breaks ties within a shared script and voices
 * Latin-script text (an English answer speaks English, a French answer with
 * French selected speaks French).
 */
export function detectSpeechLocale(text: string, preferredLang: string): string {
  const letters = text.match(/\p{L}/gu)?.length ?? 0
  if (letters > 0) {
    let bestLocale = ''
    let bestCount = 0
    for (const rule of SCRIPT_RULES) {
      const count = text.match(rule.re)?.length ?? 0
      if (count > bestCount) {
        bestCount = count
        bestLocale = rule.locale(preferredLang, text)
      }
    }
    // Dominant non-Latin script (≥25% of letters — answers often embed English
    // product names) → that script's language wins outright.
    if (bestLocale && bestCount / letters >= 0.25) return bestLocale
  }
  // Latin-script (or empty) content: the selected language if it's Latin-script,
  // otherwise it's an English answer shown while a non-Latin language is selected.
  if (!NON_LATIN_LANGS.has(preferredLang)) return SPEECH_LOCALES[preferredLang] ?? 'en-US'
  return 'en-US'
}

/* ─── Lever 2: the persona standard ─────────────────────────────────────── */

/** Time-appropriate greeting word (candidate's LOCAL part of day). */
export function greetingWord(tod?: TimeOfDay): string {
  return tod === 'morning' ? 'Good morning'
    : tod === 'afternoon' ? 'Good afternoon'
    : tod === 'evening' ? 'Good evening'
    : 'Hello'
}

/** Candidate's local part of day — client-side helper for greeting parity. */
export function localTimeOfDay(date: Date = new Date()): TimeOfDay {
  const h = date.getHours()
  return h < 12 ? 'morning' : h < 17 ? 'afternoon' : 'evening'
}

/**
 * HOW to speak — appended to every speaking persona so all surfaces share the
 * same human delivery. (The Voice Interview and the Tavus avatar both consume
 * this verbatim.)
 */
export const SPOKEN_STYLE_RULES =
  'Speak naturally and warmly, like a real person on a friendly call: use contractions, keep sentences short, one idea at a time, and vary your phrasing so you never sound templated or repetitive. ' +
  'Never read out markdown, bullet points, asterisks, code, URLs, or question numbers. Speak in plain sentences with natural pauses; do not use em dashes.'

/**
 * The varied-acknowledgment rule — a brief, warm, DIFFERENT thank-you after
 * every answer, exactly like the Voice Interview.
 */
export const VARIED_THANKS_RULE =
  'After each answer, give a brief, warm, VARIED acknowledgment before moving on — for example "Thanks, that\'s really helpful.", "Great, thank you.", "That makes sense.", "Lovely, thanks for walking me through that." — and never reuse the same phrase twice in a row.'

/** The default interviewer character used when a recruiter hasn't written one.
 *  The interviewer's NAME comes from the recruiter's avatar setup (aiName). */
export function defaultInterviewerPersona(candidateName?: string, aiName?: string): string {
  const who = candidateName?.trim() || 'the candidate'
  const me = aiName?.trim() || 'Alex'
  return `You are ${me}, a Senior Talent Specialist at TalbotIQ conducting a screening interview with ${who}. You are warm, personable and encouraging — you put people at ease and sound genuinely interested in their answers.`
}

/**
 * Build the full Tavus `conversational_context` for an interview conversation —
 * the recruiter's persona (or the default), the shared spoken style, and the
 * greet → ready-check → questions (with varied thanks) → warm wrap-up flow,
 * with the exact question script enforced. Questions are stripped for speech.
 */
export function avatarInterviewContext(opts: {
  personaText?: string
  candidateName?: string
  aiName?: string
  questions: string[]
  timeOfDay?: TimeOfDay
  /** Raw résumé text (or an excerpt) — lets the avatar sound informed about the
   *  candidate's background WITHOUT adding unscripted questions. */
  resumeText?: string
}): string {
  const who = opts.candidateName?.trim() || 'the candidate'
  const persona = opts.personaText?.trim() || defaultInterviewerPersona(opts.candidateName, opts.aiName)
  const numbered = opts.questions.map((q, i) => `${i + 1}. ${stripForSpeech(q)}`).join('\n')
  const background = opts.resumeText?.trim()
    ? `CANDIDATE BACKGROUND — from ${who}'s résumé. Use it to sound informed and to personalise your brief acknowledgments naturally (e.g. referencing their experience), but NEVER to add, change, or skip scripted questions:\n${opts.resumeText.trim().slice(0, 1500)}`
    : ''

  return [
    persona,
    ...(background ? [background] : []),
    SPOKEN_STYLE_RULES,
    `FLOW:
1. Open with a brief "${greetingWord(opts.timeOfDay)}" greeting and warmly welcome ${who}${opts.candidateName?.trim() ? ' by name' : ''}. Add one short reassuring line about how this will go, then ask if they're ready to begin, and wait.
2. If they clearly say yes, begin. If they're unsure or nervous, reassure them in one short line and ask again; only start on a clear yes.
3. Ask the questions below IN ORDER, one at a time, phrased exactly as written. Wait for ${who} to completely finish each answer — never interrupt. ${VARIED_THANKS_RULE}
4. Only AFTER the final question is answered, close warmly: thank them sincerely, tell them that's everything and they're all done, that the team will be in touch about next steps, and wish them a great rest of their day.`,
    `THE QUESTIONS, IN ORDER — ask every one, exactly as written; never say their numbers aloud:\n${numbered}`,
    `STRICT RULES: Ask ONLY these questions. Do NOT invent, add, skip, reorder, or rephrase any question, and never ask follow-ups that are not in the list. No small talk beyond the opening. If ${who} goes off-topic or asks you questions, politely acknowledge in one short line and steer straight back to the next planned question. Cover ALL the questions, then close — never finish early and never add questions of your own.`,
  ].join('\n\n')
}

/**
 * The avatar's first words. A recruiter-written greeting wins; otherwise a
 * warm, time-appropriate default that welcomes the candidate by name.
 */
export function avatarGreetingText(opts: {
  custom?: string
  candidateName?: string
  aiName?: string
  timeOfDay?: TimeOfDay
}): string {
  const custom = opts.custom?.trim()
  if (custom) return stripForSpeech(custom)
  const hi = greetingWord(opts.timeOfDay)
  const name = opts.candidateName?.trim()
  const me = opts.aiName?.trim()
  const intro = me ? `I'm ${me}, and I'm really looking forward to our chat` : `I'm really looking forward to our chat`
  return name
    ? `${hi} ${name}, welcome, and thanks so much for making the time today. ${intro}, so just relax and answer naturally. Are you ready to begin?`
    : `${hi}, welcome, and thanks so much for making the time today. ${intro}, so just relax and answer naturally. Are you ready to begin?`
}
